### PR TITLE
LibGfx/ILBM: Add an IFF-ILBM decoder :)

### PR DESCRIPTION
**Pull request by warpdesign**
IFF was a generic container fileformat that was popular on the Amiga & DOS since it was the only file format supported by Deluxe Paint. [ILBM](https://wiki.amigaos.net/wiki/ILBM_IFF_Interleaved_Bitmap) is an image format popular in the late eighties/nineties that uses the IFF container.

This is a very first version of the decoder that only supports _ByteRun_ compressed files with bpp <= 8.

Only the minimal chunks are decoded: CMAP, BODY, BMHD.

I am planning to add support for the following variants:

- EHB (32 colours + lighter 32 colours)
- HAM6 / HAM8 (special mode that allowed to display the whole Amiga 4096 colours / 262 144 colours palette)
- TrueColor (24bit)
- Mask/Transparent Color support

Things that could be fun to do:

- Still images could be animated using color cycle information

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    if (header.form != FourCC("FORM") || header.format != FourCC("ILBM"))
+        return Error::from_string_literal("Invalid IFF-ILBM header");
+
+    return {};
+}
+
+static ErrorOr<Vector<u32>> decode_cmap_chunk(Chunk cmap_chunk)
+{
+    auto data = cmap_chunk.data;
+    const u8 size = (u8)(data.size() / 3);
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:131]: We only use west `const` in this codebase :)
And avoid C-cast

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+    TRY(decode_iff_ilbm_header(*plugin->m_context));
+    TRY(decode_ilbm_first_chunk(*plugin->m_context));
+    return plugin;
+}
+
+ErrorOr<ImageFrameDescriptor> ILBMImageDecoderPlugin::frame(size_t index, Optional<IntSize>)
+{
+    if (index > 0)
+        return Error::from_string_literal("ILBMImageDecoderPlugin: frame index must be 0");
+    else {
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:354]: No need for `else` after a `return`

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        if (m_context->state < ILBMLoadingContext::ChunksDecoded)
+            TRY(decode_iff_chunks(*m_context));
+
+        RefPtr<Gfx::Bitmap> bitmap = TRY(chunky_to_bitmap(*m_context));
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:358]: That will be done every time the frame is requested, you only need to do it once

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    // points to current chunk
+    ReadonlyBytes chunks_cursor;
+
+    // max number of bytes per plane row
+    u16 pitch;
+
+    // Current file color_table in BGRA8888 format
+    Vector<u32> color_table;
+
+    ByteBuffer pixel_data;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:103]: You're basically storing the image twice, once in `pixel_data` and once in the bitmap. I'm pretty sure it can be avoided

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+
+    return color_table;
+}
+
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context)
+{
+    auto width = context.bm_header.width;
+    auto height = context.bm_header.height;
+    auto color_table = context.color_table;
+    auto chunky = context.pixel_data;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:148]: In a general manner, I'm not a fan of these aliases. I think that they hide information (in fact, where does the information come from) just to save a few characters. But that's more a question of style, so do what you want.
However, you are copying the whole bitmap here, and that's something I can argue on being bad :)

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Debug.h>
+#include <AK/Endian.h>
+#include <AK/FixedArray.h>
+#include <AK/Format.h>
+#include <AK/MemoryStream.h>
+#include <AK/String.h>
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:12]: Is this used?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2023, Nicolas Ramz <nicolas.ramz@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/MemoryStream.h>
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.h:9]: That's included both here and in the .cpp but not used

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    virtual ~ILBMImageDecoderPlugin() override;
+
+    virtual IntSize size() override;
+    virtual ErrorOr<Optional<ReadonlyBytes>> icc_data() override;
+
+    virtual ErrorOr<ImageFrameDescriptor> frame(size_t index, Optional<IntSize> ideal_size = {}) override;
+
+private:
+    ILBMImageDecoderPlugin(ReadonlyBytes, OwnPtr<ILBMLoadingContext>);
+    OwnPtr<ILBMLoadingContext> m_context;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.h:30]: Use a NonnullOwnPtr

**Review by LucasChollet — COMMENTED** [review]
Thanks for adding a decoder!

There are a lot of things, but most are trivial.

It would be really nice to add both a test and a fuzzer.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 5 above)
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:148]: It's not only to save characters but also to save some indirections: I guess that's a bad habit from the JS world. Not sure that's needed in Cpp?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+#include <LibGfx/ImageFormats/ILBMLoader.h>
+
+namespace Gfx {
+
+struct FourCC {
```
[the remark this answers is not in the document]
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:17]: Indeed: I borrowed this struct. If I want to reuse it, how am I supposed to share it? What's the habits in cpp/serenityOS to share pieces of code like this?

**Comment by warpdesign** [comment]
> Thanks for adding a decoder!
> 
> There are a lot of things, but most are trivial.
> 
> It would be really nice to add both a test and a fuzzer.

I fixed most problems: only two issues remain, see my comments.

As for the the test & fuzzer: could you point me some example as a reference?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 10]
```diff
@@ -125,27 +112,23 @@ static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
...
-    const u8 size = (u8)(data.size() / 3);
-    Vector<u32> color_table;
+    u8 const size = (data.size() / 3);
+    FixedArray<Color> color_table = TRY(FixedArray<Color>::create(size));
 
     for (int i = 0; i < size; ++i) {
-        u32 color = (data[i * 3] << 16) | (data[(i * 3) + 1] << 8) | data[(i * 3) + 2];
-        color |= 0xff000000;
-        color_table.append(color);
+        color_table[i] = Color(data[i * 3], data[(i * 3) + 1], data[(i * 3) + 2], 255);
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:122]: `Color` has a construct that only takes three parameters

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 11]
```diff
@@ -97,17 +90,11 @@ struct ILBMLoadingContext {
...
-    ByteBuffer pixel_data;
+    FixedArray<Color> color_table;
 
     RefPtr<Gfx::Bitmap> bitmap;
 
-    // BMHD IFF-ILBM header
-    BitmapHeader bm_header;
-
-    Optional<IntSize> size;
+    BitmapHeader BMHDHeader;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:97]: I was more thinking of `BMHDHeader bm_header;`
Variable names shouldn't be capitalized

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 12]
```diff
@@ -125,27 +112,23 @@ static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
     return {};
 }
 
-static ErrorOr<Vector<u32>> decode_cmap_chunk(Chunk cmap_chunk)
+static ErrorOr<FixedArray<Color>> decode_cmap_chunk(Chunk cmap_chunk)
 {
     auto data = cmap_chunk.data;
-    const u8 size = (u8)(data.size() / 3);
-    Vector<u32> color_table;
+    u8 const size = (data.size() / 3);
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:118]: Can you also remove both parentheses?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 13]
```diff
@@ -125,27 +112,23 @@ static ErrorOr<void> decode_iff_ilbm_header(ILBMLoadingContext& context)
...
-        color |= 0xff000000;
-        color_table.append(color);
+        color_table[i] = Color(data[i * 3], data[(i * 3) + 1], data[(i * 3) + 2], 255);
     }
 
     return color_table;
 }
 
-static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context)
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context, ByteBuffer& chunky)
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:128]: use a `const&` if you are only reading from it

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 14]
```diff
@@ -154,20 +137,22 @@ static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context
     for (int row = 0; row < height; ++row) {
         for (int col = 0; col < width; ++col) {
             u8 index = chunky[(width * row) + col];
-            bitmap->scanline(row)[col] = color_table[index];
+            bitmap->scanline(row)[col] = context.color_table[index].value();
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:140]: `Bitmap` has a `set_pixel` method that can take a `Color` directly

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 15]
```diff
@@ -154,20 +137,22 @@ static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context
...
 }
 
 static ErrorOr<ByteBuffer> planar_to_chunky(ReadonlyBytes bitplanes, ILBMLoadingContext& context)
 {
     dbgln_if(ILBM_DEBUG, "planar_to_chunky");
     u16 pitch = context.pitch;
-    u16 w = context.bm_header.width;
-    u16 h = context.bm_header.height;
-    u8 planes = context.bm_header.planes;
+    u16 w = context.BMHDHeader.width;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:153]: > Use full words, except in the rare case where an abbreviation would be more canonical and easier to understand.

[CodingStyle.md](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#right-1)

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 16]
```diff
@@ -225,12 +210,15 @@ static ErrorOr<void> decode_body_chunk(Chunk body_chunk, ILBMLoadingContext& con
...
+    if (context.BMHDHeader.compression == CompressionType::Cmp_ByteRun) {
         auto data = body_chunk.data;
         // these are the uncompressed interleaved bitmap planes
         auto plane_data = TRY(uncompress_byte_run(data, context));
         // that we need to convert to chunky pixel data
-        context.pixel_data = TRY(planar_to_chunky(plane_data, context));
+        auto pixel_data = TRY(planar_to_chunky(plane_data, context));
+
+        auto bitmap = TRY(chunky_to_bitmap(context, pixel_data));
+        context.bitmap = bitmap;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:221]: No need for these two steps
```cpp
context.bitmap = TRY(chunky_to_bitmap(context, pixel_data));
```

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 17]
```diff
@@ -278,14 +266,13 @@ static ErrorOr<void> decode_iff_chunks(ILBMLoadingContext& context)
         auto chunk = TRY(decode_iff_advance_chunk(chunks));
         if (chunk.type == FourCC("CMAP"))
             context.color_table = TRY(decode_cmap_chunk(chunk));
-        else if (chunk.type == FourCC("BODY"))
+        else if (chunk.type == FourCC("BODY")) {
             TRY(decode_body_chunk(chunk, context));
-        else if (chunk.type == FourCC("CRNG"))
+            context.state = ILBMLoadingContext::State::BitmapDecoded;
+        } else if (chunk.type == FourCC("CRNG"))
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:272]: https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#omission-of-curly-braces-from-statement-blocks

You're not respecting this in `decode_body_chunk`

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+    BigEndian<u16> page_height;
+};
+
+static_assert(sizeof(BitmapHeader) == 20);
+
+struct ILBMLoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+        HeaderDecoded,
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:80]: I still maintain that having both `HeaderDecoded` and `BitmapHeaderDecoded` is useless
No user should have a ILBMDecoder with its context state set to something before both the file header and the bitmap header are decoded.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:14]: Put it in a `.h`, you can probably drop it at the end of `ImageDecoder.h` so every decoder can use it.
We will move it to a utility file if we have more things to share later on.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+};
+
+static_assert(AssertSize<IFF_FILE_HEADER, 12>());
+
+struct Chunk {
+    FourCC type;
+    ReadonlyBytes data;
+};
+
+enum CompressionType : u8 {
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp]: > (that's valid everywhere)

Please do it for *every* `enum`

`enum class` do not pollute the global namespace, so you can't confuse a common name that is shared between two of them. You're manually adding a prefix to each `enum` as a workaround this issue. So please use `enum class` and drop the prefix

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 5 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp]: 1. Then you're doing premature optimization, don't try to optimize stuff without needing it and measuring it. Having a readable code is necessary, optimizing it comes after.
2. C++ compilers are very good at optimizing stuff. By doing stuff like this, you have more chance to mess with their optimizations than to outsmart them.
3. Using `auto&` would create a local reference (the size of a pointer) and that's the closest to what you want. Using `auto` and `=` calls the copy constructor which makes a __deep__ copy, copying a whole image, pixel by pixel, to save an indirection probably won't make you save some time.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+    dbgln_if(ILBM_DEBUG, "planar_to_chunky");
+    u16 pitch = context.pitch;
+    u16 w = context.bm_header.width;
+    u16 h = context.bm_header.height;
+    u8 planes = context.bm_header.planes;
+    auto chunky = TRY(ByteBuffer::create_uninitialized(w * h));
+
+    for (u16 y = 0; y < h; y++) {
+        for (u8 p = 0; p < planes; p++) {
+            const u8 plane_mask = 1 << p;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:160]: That's also an east const

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+ILBMImageDecoderPlugin::~ILBMImageDecoderPlugin() = default;
+
+IntSize ILBMImageDecoderPlugin::size()
+{
+    return IntSize { m_context->bm_header.width, m_context->bm_header.height };
+}
+
+bool ILBMImageDecoderPlugin::sniff(ReadonlyBytes data)
+{
+
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:321]: Can you remove this line?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 22]
```diff
@@ -300,23 +287,23 @@ static ErrorOr<void> decode_ilbm_first_chunk(ILBMLoadingContext& context)
...
-        context.bm_header.y,
-        context.bm_header.planes,
-        (u8)context.bm_header.mask,
-        (u8)context.bm_header.compression);
+        context.BMHDHeader.width,
+        context.BMHDHeader.height,
+        context.BMHDHeader.x,
+        context.BMHDHeader.y,
+        context.BMHDHeader.planes,
+        (u8)context.BMHDHeader.mask,
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:300]: Same as before, avoid C-cast
For enums we have `to_underlying`

**Review by LucasChollet — COMMENTED** [review]
2nd round!

When you modify your code after someone requested changes, fixup them in the corresponding commit instead of pushing a new one. (ask on discord if you face issues doing it)

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 10 above)
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:122]: I saw that, but I was already thinking about using it when I'll have implemented mask/transparent color. But I guess I can switch to this constructor when I'll be implementing that.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 16 above)
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:221]: If you think I can directly write to the bitmap, I don't think so: I really need to de-interleave the bitplanes first, then only I can write to the bitmap. So I can put both steps into the same method, but these steps can't be put into a single step.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 10 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:122]: You will need to modify the line anyway, no need to bloat it right now

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 16 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:221]: I was speaking of the two-step assignment, the temporary `bitmap` variable is useless.

```cpp
auto bitmap = TRY(chunky_to_bitmap(context, pixel_data));
context.bitmap = bitmap;
```
is just
```cpp
context.bitmap = TRY(chunky_to_bitmap(context, pixel_data));
```
but with extra steps

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp]: I put FourCC inside `ImageDecoder.h` and also modified `WebPLoader.cpp`: where FourCC was originally defined, not sure this is the way to do.

**Comment by warpdesign** [comment]
@LucasChollet I think that's it (at least for this second round :)) Do you want me to squash both commits?

**Comment by LucasChollet** [comment]
I only had a quick glance but it looks like you did not remove the prefixes from your `enum`s.

Yes, please squash them (as the English term), but use git's fixup feature.

Put the FourCC extraction in its own commit (I indeed intended you to remove it from the WebP decoder). No need to include `ImageDecoder.h` from every `*Loader.cpp` file, it's already done but including the corresponding `*Loader.h`.

**Comment by warpdesign** [comment]
> I only had a quick glance but it looks like you did not remove the prefixes from your `enum`s.
> 
> Yes, please squash them (as the English term), but use git's fixup feature.
> 
> Put the FourCC extraction in its own commit (I indeed intended you to remove it from the WebP decoder). No need to include `ImageDecoder.h` from every `*Loader.cpp` file, it's already done but including the corresponding `*Loader.h`.

Ok. I'll do it in a separate commit.

What do you mean by "remove the prefixes"?

**Comment by LucasChollet** [comment]
As an example all values in CompressionType start with "Cmp_" that's useful with a plane `enum` but useless with an `enum class` so please remove them

**Comment by warpdesign** [comment]
I moved the `FourCC` changes into a separate commit then squashed all ILBM-related ones into one commit.

I also removed the prefixes in `enum class`.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:65]: Third time's the charm, maybe?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 23]
```diff
@@ -0,0 +1,338 @@
...
+    for (int i = 0; i < size; ++i) {
+        color_table[i] = Color(data[i * 3], data[(i * 3) + 1], data[(i * 3) + 2]);
+    }
+
+    return color_table;
+}
+
+static ErrorOr<RefPtr<Gfx::Bitmap>> chunky_to_bitmap(ILBMLoadingContext& context, ByteBuffer const& chunky)
+{
+    auto& width = context.bm_header.width;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:115]: The basic rule of thumb for should I take a reference or a copy to an object is: if it fits into a register a copy is cheaper, take a reference otherwise.
Integers fall into the first category, don't take a reference if it's not needed.

Also, when I see someone taking a mutable reference, I expect it to be used to be written to. Explicitly taking a `const` reference makes your intentions much clearer.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 24]
```diff
@@ -0,0 +1,338 @@
...
+
+    return plane_data;
+}
+
+static ErrorOr<void> decode_body_chunk(Chunk body_chunk, ILBMLoadingContext& context)
+{
+    dbgln_if(ILBM_DEBUG, "decode_body_chunk {}", body_chunk.data.size());
+
+    if (context.bm_header.compression == CompressionType::ByteRun) {
+        auto& data = body_chunk.data;
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:199]: Hmm, what is this line doing?

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 25]
```diff
@@ -0,0 +1,338 @@
...
+            context.state = ILBMLoadingContext::State::BitmapDecoded;
+        } else if (chunk.type == FourCC("CRNG")) {
+            dbgln_if(ILBM_DEBUG, "Chunk:CRNG");
+        }
+    }
+
+    return {};
+}
+
+static ErrorOr<void> decode_ilbm_first_chunk(ILBMLoadingContext& context)
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:265]: If the first chunk is named BMHD, why not naming the function `decode_bmhd_chunk` (or something else if you want, just more explicit than the current name)

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 24 above)
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:199]: Removed it.

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:198]: I can still see it :eyes:

Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp]: Oops, should be really gone now.

File: Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp [hunk 26]
```diff
@@ -0,0 +1,18 @@
+/*
```
Comment (LucasChollet) [inline Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:1]: From [CONTRIBUTING](https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#code-submission-policy):
> Write the commit message subject line in the imperative mood ("Foo: Change the way dates work", not "Foo: Changed the way dates work").

And this hasn't much to do with `LibGfx`, you can make two separate commits or otherwise use `Meta+Tests` for the category.

File: Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp [hunk 27]
```diff
@@ -0,0 +1,18 @@
...
+ * Copyright (c) 2020, the SerenityOS developers.
```
Comment (LucasChollet) [inline Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:2]: I think that the year can be bumped

**Review by LucasChollet — COMMENTED** [review]
Sorry for the delay, last nitpicks, and then it should be good to go!

When you rebase your changes, make sure to keep the list in `Base/res/apps/ImageViewer.af` sorted.

File: Meta/gn/secondary/AK/BUILD.gn [hunk 28]
```diff
@@ -283,6 +283,7 @@ write_cmake_config("ak_debug_gen") {
     "HTTPJOB_DEBUG=",
     "HUNKS_DEBUG=",
     "ICO_DEBUG=",
+    "ILBM_DEBUG=",
```
[the remark this answers is not in the document]
Comment (warpdesign) [reply Meta/gn/secondary/AK/BUILD.gn:286]: I have added it in `Meta/gn/secondary/Userland/Libraries/LibGfx/BUILD.gn`: not sure if it's the only place where it's referenced.

File: Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp (hunk 26 above)
Comment (warpdesign) [reply Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:1]: I used `Meta+Tests` for the category.

Comment (LucasChollet) [reply Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:1]: You still don't use the imperative mood for this commit.
"Meta+Tests: Add a fuzzer and a test for the ILBM decoder" would be better

Comment (warpdesign) [reply Meta/Lagom/Fuzzers/FuzzILBMLoader.cpp:1]: Done :)
